### DISCUSSION TITLE
using z3 solver

### DISCUSSION THREAD

**Discussion by Nivous** [seahorn/seahorn#551] [Q&A] [answered]
Hi.
I have created a CHC system which outputs a bunch of rules and at the end a query which is a nullary pridicate fapp. I understand this is the way to use the z3 solver. I have made sure to change in the horn solver code the timeout for the fp to be unsigned(-1) which should be the maximum possible and should be plenty of time.
I am getting unknown after a really short period of time. Could you try and help? Is there anything else I need to check out?
my CHC-system contains many rules, so I am not sure how helpful it will be to see it, but for the relations and variables I defined, I will dump them right here.
Any help will be appreciated.

```
(set-info :original "../../../test_programs/test_hyper_k_basic.bc")
(set-info :authors "SeaHorn v.14.0.0-rc0")
(declare-rel hyper_bottom ())
(declare-rel hyper.doomed.state_main!_subset_0 (Int Int Int Int Int Int Int Int ))
(declare-rel hyper.doomed.state_main!_subset_0_1 (Int Int Int Int Int Int Int Int ))
(declare-rel hyper.doomed.state_main!_subset_1 (Int Int Int Int Int Int Int Int ))
(declare-rel main_1!_thread_0 (Int Int ))
(declare-rel main_1!_thread_1 (Int Int ))
(declare-rel main_1!hyper (Int Int Int Int ))
(declare-var main@%_tail2_0!_thread_0 Int )
(declare-var main@%_tail2_0!_thread_1 Int )
(declare-var main@%_tail1_1!_thread_0 Int )
(declare-var main@%_tail1_1!_thread_1 Int )
(declare-var main@%_tail1_0!_thread_0 Int )
(declare-var main@%_tail1_0!_thread_1 Int ) 
```

**Comment by Nivous** [comment]
Any help would be appreciated

**Comment by agurfinkel** [comment]
There is not much I can say based on the information provided. You are showing that there are number of declarations. Thes clarations seem fine. 

Perhaps you can send the fp file that you generate. Usually these are sent to us through some file sharing service, like github gist.

**Comment by Nivous** [comment]
@agurfinkel 
Thank you
I am trying to implement the reduction @yvizel published in his article [here](https://arxiv.org/pdf/2304.12588.pdf)
I have created my own fork for that purpose, and tried using the code I wrote on the following program:

```
#include "seahorn/seahorn.h"

extern int nd(void);

int main(void) {
    int a = nd();

    __hyper_pre_gt(a);
	
	a = a + 1;

    __hyper_post_gt(a);
	
    return 0;
}
```

which basically just check what is obvious - that is if we get an input a which is larger than input b it should still be the case at the end of the function.

From the modified seahorn code I got the following smt file: https://gist.github.com/Nivous/50c9d22edc00b703e1ec27178968df98

which yields an unknown result from z3 when using --solve

I was hoping you could help me debug \ figure out what went wrong.

**Comment by agurfinkel** [comment, marked as answer]
When I run z3 on your instance, it prints an error message:
```
(error "query failed: Rule contains negative predicate <null>:
P!!0(#0,#1) :-
 not main_1!_thread_1(3,#0).
")
```

after the error, the result itself is irrelevant since there was an error in the input.

Looking at your input, you seem to have predicate in negative positions in a rule. This is not allowed in CHCs.
Specifically, this rule seems to be the problem:

```
(rule (let ((a!1 (or (and (not (main_1!_thread_0 3 main@%_tail1_0!_thread_0))
                    (not (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))
               (and (main_1!_thread_0 3 main@%_tail1_0!_thread_0)
                    (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))))
  (=> (not a!1)
      (hyper.doomed.state_main!_subset_0_1
        0
        0
        main@%_tail2_0!_thread_0
        main@%_tail2_0!_thread_1
        main@%_tail1_1!_thread_0
        main@%_tail1_1!_thread_1
        main@%_tail1_0!_thread_0
        main@%_tail1_0!_thread_1))))
```
`a!1` is negated, so all application of predicates under `a!1` must be negated as well. Some are, but some are not. Must be a mistake in how you generate the constraints.

**Reply by Nivous** [reply]
Thank you very much for the help.
I took a look at that and how I create this rule. I don't mention this a!1. It is generated by z3 (I think) somehow. If we ignore the a!1 statements this rule Is what I meant to create. Meaning:

```
(rule (let ((a!1 (or (and (not (main_1!_thread_0 3 main@%_tail1_0!_thread_0))
                    (not (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))
               (and (main_1!_thread_0 3 main@%_tail1_0!_thread_0)
                    (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))))
  (=> (not a!1)
      (hyper.doomed.state_main!_subset_0_1
        0
        0
        main@%_tail2_0!_thread_0
        main@%_tail2_0!_thread_1
        main@%_tail1_1!_thread_0
        main@%_tail1_1!_thread_1
        main@%_tail1_0!_thread_0
        main@%_tail1_0!_thread_1))))
```

And from looking at my code it does seem to be what I am creating. But these a!1 statements are generated non the less.
also, when removing these rules the solver seems to run without any errors, but also no a!1 statements are created.
Is there something I can do to avoid having these statements or at least figure out when are they inserted into the code?

**Comment by agurfinkel** [comment]
`a!1` is not a problem. They are just `let` bindings that are created by `z3` to improve printing. In the solver, each occurrence of `a!1` is replaced by its definition.

The syntax of the `let` statement is roughly
```
(let ( (name1 def1) (name2 def2) ... ) body)
```
where `name1` is a name, often `a!1`, `def1` is a formula that is a definition for `a!1` and `body` is a formula that can use `name1`.

You can always manually inline `def_i` for each `name_i`.

**Comment by Nivous** [comment]
@agurfinkel 
Again, thanks for the reply, it is really helpful.
I see your point. Then you are saying let statements are basically like creating a macro for some rule?
Also, I assume the definition only holds for the body? And when "exiting" the let parenthesis it is no longer defined?
If so,  then I am not sure what is the problem with this clause? (the one you mentioned is faulty in your first reply):

```
(rule (let ((a!1 (or (and (not (main_1!_thread_0 3 main@%_tail1_0!_thread_0))
                    (not (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))
               (and (main_1!_thread_0 3 main@%_tail1_0!_thread_0)
                    (main_1!_thread_1 3 main@%_tail1_0!_thread_1)))))
  (=> (not a!1)
      (hyper.doomed.state_main!_subset_0_1
        0
        0
        main@%_tail2_0!_thread_0
        main@%_tail2_0!_thread_1
        main@%_tail1_1!_thread_0
        main@%_tail1_1!_thread_1
        main@%_tail1_0!_thread_0
        main@%_tail1_0!_thread_1))))
```
